### ISSUE TITLE
Identify exact pokemon by candy name + evolution cost

### ISSUE THREAD

**Issue by nahojjjen** [farkam135/GoIV#184]
Would require an xml listing all evolution costs, but in my opinion has the potential to be a better recognition method than trying to determine by nickname when pokemon has been renamed.

**Comment by Blaisorblade** [comment]
Here's a source list (not double-checked, taken from http://gaming.stackexchange.com/a/274429/66181): http://www.serebii.net/pokemongo/evolution.shtml

I agree this sounds good—except for OCR errors, this method *always* gives the correct solution. To me the main question is If OCR works well on that field (and if we can fix candy misdetections like #182).

**Comment by sarav** [comment]
Actually, I'm pretty sure this won't work. The power up cost doesn't change across evolutions. I have plenty of examples because I used to record some evolutions. The cost is the same before and after evolution and evolution doesn't change levels.

**Comment by Blaisorblade** [comment]
To clarify, since *I* keep mixing this up: this is the number of candies needed for *evolution*, not the stardust and candies needed for *powerup* (which is constant across levels).

**Comment by sarav** [comment]
I had the same confusion, made a comment and deleted it.

**Comment by sarav** [comment]
But this means we also have to detect the evolve section not being there.

**Comment by Blaisorblade** [comment]
> But this means we also have to detect the evolve section not being there.

Yes, but if somebody can guess the magic numbers to find the evolve section (which I can't*), detecting the evolve section is comparatively easy enough, you find either white background or an horizontal separator rather than text on a grey background. The code for this is dark magic, but the one we have is working well enough...

*Actually, it's more that I have no real way to be confident I got this right.

**Comment by DJCrapsody** [comment]
I haven't wrapped my head around on the method the ocr was programmed to detect Nidoran color differences but we could do the same way. Just to check whether its whitish or greenish.

**Comment by mattaereal** [comment]
Hi guys, I just read this very quickly. A few ideas. For general purpose pokemons (not thinking about Eevee and it's future evolutions neither magikarp, this cases could be arranged later) I'm picturing a HashMap. One entry for every candy name. Then, inside of each entry, an array of strings. First position the evolution for 12 candys, second position the evolution for 25, third 50, forth 100. 

Also, another hashmap, that given 12 gives us 0, 1 for 25, 2 for 50, 3 for 100.

It would be something like this:


```
evolutionMap:
[bulbasaur] -> {null, ivysaur, null, venosaur},
{charmander} -> {null, charmeleon, null, charizard},
...
{pidgey} -> {pidgeotto, null, pidgeot, null},
{horsea} -> {null, null, seadra, null},
...
```

Then the map, enum, or whatever you imagine has the following values, let's imagine it's map and call it indexMap for a sec:
```
12 -> 0
25 -> 1
50 -> 2
100 -> 3
```

The core implementation could be something like this:
```
String getPokemon(String candyName, int candyAmount) {
    return evolutionMap.get(candyName)[indexMap.get(candyAmount)];
}
```
No complexity at all, and we just need to add a case for Eeevee and Magikarp. Also, the values from the evolutionMap could easily be stored as a json format, or an XML.

**Comment by sarav** [comment]
Yeah, I don't think getting the pokemon from candy name and candy count is the hard part. It's finding and putting in that data in the resource files and OCRing these things that's hard.

Btw, we don't need all these new hash maps, etc because it's kinda there already (pokemap). We just need to add a "candynum" to each pokemon that has an evolution. Then, it's just a matter of finding the base pokemon with candy name (there's already a get(name) for this) and the following the evolutions to find the right one (pokemon.candynum == what we read).

**Comment by Paxtez** [comment]
To add on to this, could use something similar to the Nido color detection to detect the background color for the eevee evolutions, to determine which one is selected.

**Comment by nahojjjen** [comment]
I have begun working on this in the branch https://github.com/farkam135/GoIV/tree/candyOnlyIdentification

**Comment by sarav** [comment]
Whatever happened to this? Click the link gives me 404 error.

**Comment by nahojjjen** [comment]
I have added the evolution candy cost information for each pokemon, so the required data is there. I havent been involved with the code that weights what information is available and spits out a pokemon name, and I had other things to do, so I dropped it for the time being. 

What needs to be done is:
The candy number needs to be OCR-d
The OCR'd info needs to be included in the pokemon-picking algorithm.

**Comment by Blaisorblade** [comment]
Anybody else plans to work on this for 3.1? Scheduling for 3.3 for now, as it's not a bugfix.

**Comment by sarav** [comment]
If someone can handle getting the data into populatePokemoninfo() and updated OCRHelper to extract the value (I still have no idea how the numbers are figured out) I can do that rest.

There's not much left at that point, but if others feel they aren't familiar with the rest of the logic, I can do that part.

**Comment by Blaisorblade** [comment]
> updated OCRHelper to extract the value (I still have no idea how the numbers are figured out)

Me neither—IIRC @farkam135 asked for lots of screenshots originally.
For the backend logic, you wrote the original code and it's not short, so that sounds good. We might want to add tests there though.

**Comment by nahojjjen** [comment]
I have started implementing this in 250628596279f40292ee93fe2b2f31393987f106

**Comment by sarav** [comment]
@nahojjjen Let me know if you want me to help out with the stuff I offer to help out on.

**Comment by nahojjjen** [comment]
Was merged in #502
